File: tests/Explorer.Core.Tests/ViewModels/PaneViewModelTests.cs
using NoctusExplorer.Core.Abstractions;
using NoctusExplorer.Core.Models;
using NoctusExplorer.Core.ViewModels;
using NSubstitute;
using FluentAssertions;

namespace NoctusExplorer.Core.Tests.ViewModels;

public class PaneViewModelTests
{
    private static PaneViewModel CreatePane()
    {
        var shell = Substitute.For<IShellService>();
        var watcher = Substitute.For<IFileWatcher>();
        shell.GetSpecialFolder(SpecialFolder.Home).Returns(new PathRef("/home", isDirectory: true));
        shell.GetDisplayName(Arg.Any<PathRef>()).Returns(ci => ci.Arg<PathRef>().DisplayName);
        return new PaneViewModel(shell, watcher);
    }

    [Fact]
    public void Initially_NoTabs()
    {
        var pane = CreatePane();
        pane.Tabs.Should().BeEmpty();
        pane.ActiveTab.Should().BeNull();
    }

    [Fact]
    public void AddTab_CreatesTabAndActivatesIt()
    {
        var pane = CreatePane();
        var tab = pane.AddTab();
        pane.Tabs.Should().HaveCount(1);
        pane.ActiveTab.Should().Be(tab);
    }

    [Fact]
    public void AddTab_WithLocation_SetsTabLocation()
    {
        var pane = CreatePane();
        var loc = new PathRef("/docs", isDirectory: true);
        var tab = pane.AddTab(loc);
        tab.Location.Should().Be(loc);
    }

    [Fact]
    public void CloseTab_RemovesTab()
    {
        var pane = CreatePane();
        var tab = pane.AddTab();
        pane.CloseTab(tab.Id);
        pane.Tabs.Should().BeEmpty();
    }

    [Fact]
    public void CloseTab_ActiveTab_ActivatesNeighbor()
    {
        var pane = CreatePane();
        var tab1 = pane.AddTab();
        var tab2 = pane.AddTab();
        pane.CloseTab(tab2.Id);
        pane.ActiveTab.Should().Be(tab1);
    }

    [Fact]
    public void ActivateTab_SwitchesActiveTab()
    {
        var pane = CreatePane();
        var tab1 = pane.AddTab();
        var tab2 = pane.AddTab();
        pane.ActivateTab(tab1.Id);
        pane.ActiveTab.Should().Be(tab1);
    }

    [Fact]
    public void RestoreTabs_PopulatesFromState()
    {
        var pane = CreatePane();
        var states = new List<TabState>
        {
            new(0, new PathRef("/a", isDirectory: true)),
            new(1, new PathRef("/b", isDirectory: true), ViewMode.List)
        };
        pane.RestoreTabs(states);
        pane.Tabs.Should().HaveCount(2);
        pane.Tabs[1].ViewMode.Should().Be(ViewMode.List);
    }
}

File: macos/Tests/NoctusCoreTests/PaneViewModelTests.swift
import Testing
@testable import NoctusCore

@Suite struct PaneViewModelTests {

    private func createPane() -> PaneViewModel {
        let shell = MockShellService()
        let watcher = MockFileWatcher()
        return PaneViewModel(shellService: shell, fileWatcher: watcher)
    }

    @Test func initiallyNoTabs() {
        let pane = createPane()
        #expect(pane.tabs.isEmpty)
        #expect(pane.activeTab == nil)
    }

    @Test func addTabCreatesTabAndActivatesIt() {
        let pane = createPane()
        let tab = pane.addTab()
        #expect(pane.tabs.count == 1)
        #expect(pane.activeTab?.id == tab.id)
    }

    @Test func addTabWithLocationSetsTabLocation() {
        let pane = createPane()
        let loc = PathRef("/docs", isDirectory: true)
        let tab = pane.addTab(at: loc)
        #expect(tab.location == loc)
    }

    @Test func closeTabRemovesTab() {
        let pane = createPane()
        let tab = pane.addTab()
        pane.closeTab(tab.id)
        #expect(pane.tabs.isEmpty)
    }

    @Test func closeActiveTabActivatesNeighbor() {
        let pane = createPane()
        let tab1 = pane.addTab()
        let tab2 = pane.addTab()
        pane.closeTab(tab2.id)
        #expect(pane.activeTab?.id == tab1.id)
    }

    @Test func activateTabSwitchesActiveTab() {
        let pane = createPane()
        let tab1 = pane.addTab()
        _ = pane.addTab()
        pane.activateTab(tab1.id)
        #expect(pane.activeTab?.id == tab1.id)
    }

    @Test func restoreTabsPopulatesFromState() {
        let pane = createPane()
        let states = [
            TabState(id: 0, location: PathRef("/a", isDirectory: true)),
            TabState(id: 1, location: PathRef("/b", isDirectory: true), viewMode: .list)
        ]
        pane.restoreTabs(states)
        #expect(pane.tabs.count == 2)
        #expect(pane.tabs[1].viewMode == .list)
    }
}
